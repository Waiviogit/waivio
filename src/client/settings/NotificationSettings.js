import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { isEqual } from 'lodash';
import { injectIntl, FormattedMessage } from 'react-intl';
import { Checkbox, Form, InputNumber, message } from 'antd';
import { updateProfile, reload } from '../../store/authStore/authActions';
import withEditor from '../components/Editor/withEditor';
import Action from '../components/Button/Action';
import requiresLogin from '../auth/requiresLogin';
import { saveNotificationsSettings } from '../../common/helpers/metadata';
import { notificationType } from '../../common/constants/waivio';
import { updateUserMetadata } from '../../store/usersStore/usersActions';
import {
  getAuthenticatedUserName,
  getAuthenticatedUserNotificationsSettings,
} from '../../store/authStore/authSelectors';

import './Settings.less';

@requiresLogin
@injectIntl
@connect(
  state => ({
    userName: getAuthenticatedUserName(state),
    settingsNotifications: getAuthenticatedUserNotificationsSettings(state),
  }),
  {
    updateProfile,
    reload,
    updateUserMetadata,
  },
)
@withEditor
export default class NotificationSettings extends React.Component {
  static propTypes = {
    intl: PropTypes.shape().isRequired,
    settingsNotifications: PropTypes.shape().isRequired,
    userName: PropTypes.string,
    updateUserMetadata: PropTypes.func.isRequired,
  };

  static defaultProps = {
    userName: '',
  };

  constructor(props) {
    super(props);
    this.state = {
      notifications: {
        activationCampaign: props.settingsNotifications.activationCampaign,
        follow: props.settingsNotifications.follow,
        fillOrder: props.settingsNotifications.fillOrder,
        mention: props.settingsNotifications.mention,
        minimalTransfer: props.settingsNotifications.minimalTransfer,
        reblog: props.settingsNotifications.reblog,
        reply: props.settingsNotifications.reply,
        statusChange: props.settingsNotifications.statusChange,
        transfer: props.settingsNotifications.transfer,
        witness_vote: props.settingsNotifications.witness_vote,
        myPost: props.settingsNotifications.myPost,
        myComment: props.settingsNotifications.myComment,
        myLike: props.settingsNotifications.myLike,
        like: props.settingsNotifications.like,
        downvote: props.settingsNotifications.downvote,
        claimReward: props.settingsNotifications.claimReward,
        powerUp: props.settingsNotifications.powerUp,
      },

      isLoading: false,
    };
  }

  handleSubmit = e => {
    e.preventDefault();
    this.setState({ isLoading: true });
    saveNotificationsSettings(this.state.notifications, this.props.userName)
      .then(res => {
        this.setState({ isLoading: false });
        this.props.updateUserMetadata(res);
        message.success(this.props.intl.formatMessage({ id: 'saved', defaultMessage: 'Saved' }));
      })
      .catch(err => message.error(err.message));
  };

  onChangeCheckbox = fields =>
    this.setState(prevState => ({
      ...prevState,
      notifications: {
        ...prevState.notifications,
        [fields]: !prevState.notifications[fields],
      },
    }));

  render() {
    const { notifications } = this.state;

    return (
      <div className="center">
        <h1>
          <FormattedMessage id="notification_settings" defaultMessage="Notification Settings" />
        </h1>
        <Form onSubmit={this.handleSubmit}>
          <div className="Settings">
            <div className="Settings__section">
              <h3>
                <FormattedMessage
                  id="instant_mobile_notifications"
                  defaultMessage="Instant mobile notifications:"
                />
              </h3>
              <p className="mt3">
                <FormattedMessage
                  id="instant_mobile_notifications_description"
                  defaultMessage="You can now receive instant mobile notifications via the Telegram app when someone replies to or re-blogs your post on Hive, mentions you, follows you, transfers funds to you, and so on."
                />
              </p>
              <p className="mt3">
                <FormattedMessage
                  id="instant_mobile_notifications_open"
                  defaultMessage="Open the Telegram chart with"
                />
                <a href="http://t.me/waivio_notifications_bot" target="_blank" rel="noreferrer">
                  @waivio_notifications_bot
                </a>
                <FormattedMessage
                  id="instant_mobile_notifications_hive"
                  defaultMessage="and enter the Hive usernames to subscribe."
                />
              </p>
            </div>
            <div className="Settings__section">
              <h3>
                <FormattedMessage id="сommunity_actions" defaultMessage="Community actions" />:
              </h3>
              {notificationType.сommunityActions.map(notify => (
                <div className="Settings__section__checkbox" key={notify.name}>
                  <Checkbox
                    checked={notifications[notify.name]}
                    onChange={() => this.onChangeCheckbox(notify.name)}
                  >
                    <FormattedMessage id={notify.id} defaultMessage={notify.defaultMessage} />
                  </Checkbox>
                </div>
              ))}
            </div>
            <div className="Settings__section">
              <h3>
                <FormattedMessage id="wallet_transactions" defaultMessage="Wallet transactions" />:
              </h3>
              <div className="Settings__section__checkbox Settings__section--flex-wrapper">
                <Checkbox
                  checked={notifications.transfer}
                  onChange={() => this.onChangeCheckbox('transfer')}
                >
                  <FormattedMessage id="incoming_transfers" defaultMessage="Incoming transfers" />
                </Checkbox>
                (
                <FormattedMessage
                  id="min_amount"
                  defaultMessage="min. amount: {input} USD"
                  values={{
                    input: (
                      <InputNumber
                        step={0.1}
                        min={0}
                        size="small"
                        defaultValue={notifications.minimalTransfer}
                        onChange={e =>
                          this.setState(prevState => ({
                            ...prevState,
                            notifications: {
                              ...prevState.notifications,
                              minimalTransfer: e,
                            },
                          }))
                        }
                      />
                    ),
                  }}
                />
                )
              </div>
              {notificationType.walletTransactions.map(notify => (
                <div className="Settings__section__checkbox" key={notify.name}>
                  <Checkbox
                    checked={notifications[notify.name]}
                    onChange={() => this.onChangeCheckbox(notify.name)}
                  >
                    <FormattedMessage id={notify.id} defaultMessage={notify.defaultMessage} />
                  </Checkbox>
                </div>
              ))}
            </div>
            <div className="Settings__section">
              <h3>
                <FormattedMessage id="my_actions" defaultMessage="My actions" />:
              </h3>
              {notificationType.myActions.map(notify => (
                <div className="Settings__section__checkbox" key={notify.name}>
                  <Checkbox
                    checked={notifications[notify.name]}
                    onChange={() => this.onChangeCheckbox(notify.name)}
                  >
                    <FormattedMessage id={notify.id} defaultMessage={notify.defaultMessage} />
                  </Checkbox>
                </div>
              ))}
            </div>
            <div className="Settings__section">
              <h3>
                <FormattedMessage
                  id="security_alerts"
                  defaultMessage="Security alerts (always on)"
                />
                :
              </h3>
              <div className="Settings__section__inputs">
                <div className="Settings__section__checkbox">
                  <Checkbox checked disabled>
                    <FormattedMessage
                      id="security_alerts_info"
                      defaultMessage="Outgoing transfers, power downs, withdrawals from savings, delegations, password recovery requests, recovery address changes, app authorizations"
                    />
                  </Checkbox>
                </div>
              </div>
            </div>
            <Action
              primary
              big
              type="submit"
              disabled={isEqual(this.props.settingsNotifications, notifications)}
              loading={this.state.isLoading}
            >
              <FormattedMessage id="save" defaultMessage="Save" />
            </Action>
          </div>
        </Form>
      </div>
    );
  }
}
