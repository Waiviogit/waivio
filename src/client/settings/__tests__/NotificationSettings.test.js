import React from 'react';
import { shallow } from 'enzyme';
import { Provider } from 'react-redux';
import { IntlProvider } from 'react-intl';
import configureStore from 'redux-mock-store';
import NotificationSettings from '../NotificationSettings';

describe('(Component) NotificationSettings', () => {
  it('renders and matches snapshot', () => {
    const mockStore = configureStore();
    const props = {
      settingsNotifications: {},
      updateUserMetadata: () => {},
    };
    const wrapper = shallow(
      <Provider store={mockStore()}>
        <IntlProvider locale="en">
          <NotificationSettings {...props} />
        </IntlProvider>
      </Provider>,
    );
    expect(wrapper).toMatchSnapshot();
  });
});
