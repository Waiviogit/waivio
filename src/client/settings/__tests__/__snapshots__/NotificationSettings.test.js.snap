// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Component) NotificationSettings renders and matches snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Provider
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  >
    <IntlProvider
      locale="en"
    >
      <UNDEFINED
        settingsNotifications={Object {}}
        updateUserMetadata={[Function]}
      />
    </IntlProvider>
  </Provider>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": <IntlProvider
        locale="en"
      >
        <UNDEFINED
          settingsNotifications={Object {}}
          updateUserMetadata={[Function]}
        />
      </IntlProvider>,
      "value": Object {
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "subscription": Subscription {
          "handleChangeWrapper": [Function],
          "listeners": Object {
            "notify": [Function],
          },
          "onStateChange": [Function],
          "parentSub": undefined,
          "store": Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          },
          "unsubscribe": null,
        },
      },
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <UNDEFINED
          settingsNotifications={Object {}}
          updateUserMetadata={[Function]}
        />,
        "locale": "en",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "settingsNotifications": Object {},
          "updateUserMetadata": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": Object {
          "$$typeof": Symbol(react.memo),
          "WrappedComponent": [Function],
          "compare": null,
          "displayName": "Connect(RequiresLogin(InjectIntl(Connect(Connect(InjectIntl(withEditor(NotificationSettings)))))))",
          "type": [Function],
          Symbol(Symbol.hasInstance): [Function],
        },
      },
      "type": [Function],
    },
    "type": Object {
      "$$typeof": Symbol(react.provider),
      "_context": Object {
        "$$typeof": Symbol(react.context),
        "Consumer": Object {
          "$$typeof": Symbol(react.context),
          "_calculateChangedBits": null,
          "_context": [Circular],
        },
        "Provider": [Circular],
        "_calculateChangedBits": null,
        "_currentRenderer": null,
        "_currentRenderer2": null,
        "_currentValue": null,
        "_currentValue2": null,
        "_threadCount": 0,
        "displayName": "ReactRedux",
      },
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <IntlProvider
          locale="en"
        >
          <UNDEFINED
            settingsNotifications={Object {}}
            updateUserMetadata={[Function]}
          />
        </IntlProvider>,
        "value": Object {
          "store": Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          },
          "subscription": Subscription {
            "handleChangeWrapper": [Function],
            "listeners": Object {
              "notify": [Function],
            },
            "onStateChange": [Function],
            "parentSub": undefined,
            "store": Object {
              "clearActions": [Function],
              "dispatch": [Function],
              "getActions": [Function],
              "getState": [Function],
              "replaceReducer": [Function],
              "subscribe": [Function],
            },
            "unsubscribe": null,
          },
        },
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <UNDEFINED
            settingsNotifications={Object {}}
            updateUserMetadata={[Function]}
          />,
          "locale": "en",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "settingsNotifications": Object {},
            "updateUserMetadata": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": Object {
            "$$typeof": Symbol(react.memo),
            "WrappedComponent": [Function],
            "compare": null,
            "displayName": "Connect(RequiresLogin(InjectIntl(Connect(Connect(InjectIntl(withEditor(NotificationSettings)))))))",
            "type": [Function],
            Symbol(Symbol.hasInstance): [Function],
          },
        },
        "type": [Function],
      },
      "type": Object {
        "$$typeof": Symbol(react.provider),
        "_context": Object {
          "$$typeof": Symbol(react.context),
          "Consumer": Object {
            "$$typeof": Symbol(react.context),
            "_calculateChangedBits": null,
            "_context": [Circular],
          },
          "Provider": [Circular],
          "_calculateChangedBits": null,
          "_currentRenderer": null,
          "_currentRenderer2": null,
          "_currentValue": null,
          "_currentValue2": null,
          "_threadCount": 0,
          "displayName": "ReactRedux",
        },
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
}
`;
